pendulum_control/CMakeLists.txt

```diff
@@ -66,6 +66,45 @@ install(
 if(AMENT_ENABLE_TESTING)
   find_package(ament_lint_auto REQUIRED)
   ament_lint_auto_find_test_dependencies()
+
+  if(NOT WIN32)
+    set(test_executable_extension "")
+    set(test_executable_subfolder "")
+  else()
+    set(test_executable_extension ".exe")
+    set(test_executable_subfolder "/${CMAKE_BUILD_TYPE}")
+    if("${test_executable_subfolder}" STREQUAL "/")
+      set(test_executable_subfolder "/Debug")
+    endif()
+  endif()
+
+  # get the rmw implementations ahead of time
+  get_available_rmw_implementations(middleware_implementations)
+  foreach(middleware_impl ${middleware_implementations})
+    find_package("${middleware_impl}" REQUIRED)
+  endforeach()
+
+  set(RCLCPP_DEMO_PENDULUM_LOGGER_EXPECTED_OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/test/pendulum_logger")
+  set(RCLCPP_DEMO_PENDULUM_DEMO_EXPECTED_OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/test/pendulum_demo")
+  set(RCLCPP_DEMO_PENDULUM_TELEOP_EXPECTED_OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/test/pendulum_teleop")
+
+  foreach(middleware_impl ${middleware_implementations})
+    set(RCLCPP_DEMO_PENDULUM_LOGGER_EXECUTABLE "${CMAKE_CURRENT_BINARY_DIR}${test_executable_subfolder}/pendulum_logger__${middleware_impl}${test_executable_extension}")
+    set(RCLCPP_DEMO_PENDULUM_DEMO_EXECUTABLE "${CMAKE_CURRENT_BINARY_DIR}${test_executable_subfolder}/pendulum_demo__${middleware_impl}${test_executable_extension}")
+    set(RCLCPP_DEMO_PENDULUM_TELEOP_EXECUTABLE "${CMAKE_CURRENT_BINARY_DIR}${test_executable_subfolder}/pendulum_teleop__${middleware_impl}${test_executable_extension}")
+
+    configure_file(
+      test/test_pendulum_demo.py.in
+      test_pendulum__${middleware_impl}.py
+      @ONLY
+    )
+
+    ament_add_nose_test(test_demo_pendulum__${middleware_impl}
+      "${CMAKE_CURRENT_BINARY_DIR}/test_pendulum__${middleware_impl}.py"
+      TIMEOUT 30)
+    set_tests_properties(test_demo_pendulum__${middleware_impl}
+    PROPERTIES DEPENDS "test_demo_pendulum__${middleware_impl} test_demo_pendulum__${middleware_impl}")
+  endforeach()
```

Comment:
You might want to use the new CMake function `call_for_each_rmw_implementation` for this (https://github.com/ros2/rmw/pull/50).